lib/libzfs/libzfs_util.c

```diff
@@ -661,26 +686,80 @@ libzfs_run_process(const char *path, char *argv[], int flags)
 	return (-1);
 }
 
-int
+/*
+ * Verify the required ZFS_DEV device is available and optionally attempt
+ * to load the ZFS modules.  Under normal circumstances the modules
+ * should already have been loaded by some external mechanism.
+ *
+ * Environment variables:
+ * - ZFS_MODULE_LOADING="YES|yes|ON|on" - Attempt to load modules.
+ * - ZFS_MODULE_TIMEOUT="<seconds>"     - Seconds to wait for ZFS_DEV
+ */
+static int
 libzfs_load_module(const char *module)
 {
 	char *argv[4] = {"/sbin/modprobe", "-q", (char *)module, (char *)0};
+	char *load_str, *timeout_str;
+	long timeout = 10; /* seconds */
+	long busy_timeout = 10; /* milliseconds */
+	int load = 0, fd;
+	hrtime_t start;
+
+	/* Optionally request module loading */
+	if (!libzfs_module_loaded(module)) {
+		load_str = getenv("ZFS_MODULE_LOADING");
+		if (load_str &&
+		    (!strncasecmp(load_str, "YES", strlen("YES")) ||
+		    (!strncasecmp(load_str, "ON", strlen("ON")))))
+			load = 1;
+
+		if (load && libzfs_run_process("/sbin/modprobe", argv, 0))
+			return (ENOEXEC);
+	}
 
-	if (libzfs_module_loaded(module))
-		return (0);
+	/* Module loading is synchronous it must be available */
+	if (!libzfs_module_loaded(module))
```

Comment:
Has anyone checked to see if the kernel code will make the entries in /sys before returning to userspace? If that is not the case, there is still a race here (although smaller).